FitGoal/FitGoal/Singup&Singin/SocialMediaAuthenticator.swift

```diff
@@ -53,4 +69,127 @@ class SocialMediaAuthenticator: NSObject, GIDSignInDelegate {
+    
+    //MARK: - Facebook
+    
+    private func facebookSignIn(sender: UIViewController, completion: @escaping SignInCallback) {
+        let permissions = ["public_profile", "email"]
+        LoginManager().logIn(permissions: permissions, from: sender) { (logInResuts, error) in
+            if let error = error {
+                completion(.failure(error))
+            }
+            else {
+                guard let logInResuts = logInResuts else {
+                    let error = LoginError.noLogInResultsFound
+                    completion(.failure(error))
+                    return
+                }
+                
+                if logInResuts.isCancelled {
+                    let error = LoginError.userCanceledLogIn
+                    completion(.failure(error))
+                }
+                else {
+                    guard let user = Auth.auth().currentUser else {
+                        let error = MissingUserInfoError.noUserFound
+                        completion(.failure(error))
+                        return
+                    }
+                    
+                    guard let email = user.email else {
+                        let error = MissingUserInfoError.noEmailFound
+                        completion(.failure(error))
+                        return
+                    }
+                    
+                    Auth.auth().fetchSignInMethods(forEmail: email) { (signInMethods, error) in
+                        if let error = error {
+                            completion(.failure(error))
+                        }
+                        else if signInMethods != nil && !signInMethods!.contains("facebook.com") {
+                            self.mergeNewUserWith(currentUser: user, completion: completion)
+                        } else {
+                            self.createNewUser(completion: completion)
+                        }
+                    }
+                }
+            }
+        }
+    }
+    
+    private func mergeNewUserWith(currentUser: User, completion: @escaping SignInCallback) {
+        do {
+            let credentials = try getFacebookCredentials()
+            currentUser.link(with: credentials) { (dataResults, error) in
+                self.manageResutls(dataResults: dataResults, error: error, completion: completion)
+            }
+        }
+        catch {
+            completion(.failure(error))
+        }
+    }
+    
+    private func createNewUser(completion: @escaping SignInCallback) {
+        do {
+            let credentials = try getFacebookCredentials()
+            Auth.auth().signIn(with: credentials) { ( dataResults, error) in
+                self.manageResutls(dataResults: dataResults, error: error, completion: completion)
+            }
+        } catch {
+            completion(.failure(error))
+        }
+    }
+    /*
+     I've moved this part of the logic to an independent func to avoid repeating code. Not sure if it's the best way
```

Comment:
I think it's great to have this logic separated here. I think the naming could be improved but in the current form is ok.